### PR TITLE
First extension arch

### PR DESCRIPTION
**Pull request by tbille**

### CODE REVIEW

File: canonicalwebteam/blog_extension/app.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from webapp.blog_ext.views import build_blueprint
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/app.py]: Did you leave this in by accident? I'm not sure calling a function outside this module, that you expect to exist inside the wrapping webapp is a good idea.

(also not crazy about a blueprint builder - seems kinda obfuscating)

File: canonicalwebteam/blog_extension/logic.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+def change_url(feed, host):
+    """Change insights urls to snapcraft.io/blog
+
+    :param feed: String with urls
+
+    :returns: A string with converted urls
+    """
+    url_regex = re.compile(
+        "https:\/\/admin.insights.ubuntu.com(\/\d{4}\/\d{2}\/\d{2})?"
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/logic.py]: As mentioned in snapcraft.io, I suspect you don't need to escape the slashes here.

File: canonicalwebteam/blog_extension/logic.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    :returns: Boolean
+    """
+    for tag in tags:
+        if tag["name"].startswith("sc:series"):
+            return True
+
+    return False
+
+
+def whitelist_categories(categories):
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/logic.py]: Is this whitelist definitely the same for needed for Japan? Why do we need a whitelist at all?

File: canonicalwebteam/blog_extension/logic.py (hunk 2 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/logic.py]: yeah I was wainting to fix on s.io before applying the same configuration

File: canonicalwebteam/blog_extension/app.py (hunk 1 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/app.py]: yes this was by testing on s.io I need to remove it

File: canonicalwebteam/blog_extension/logic.py (hunk 3 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/logic.py]: this can be removed

File: canonicalwebteam/blog_extension/views.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            ),
+            None,
+        )
+
+        try:
+            articles, total_pages = api.get_articles(
+                tags=tags_id, page=page_param, category=filter_category
+            )
+        except Exception as e:
+            print(e)
```
Comment (b-m-f) [inline canonicalwebteam/blog_extension/views.py]: Use a logger here instead?

File: canonicalwebteam/blog_extension/views.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
+import flask
+
+from webapp.blog_ext import api
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/views.py]: again with the webapp

File: canonicalwebteam/blog_extension/logic.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    """Prefixes images with cloudinary optimised URLs and adds srcset for
+    image scaling
+
+    :param content: The HTML string to convert
+
+    :returns: Update HTML string with converted images
+    """
+    cloudinary = "https://res.cloudinary.com/"
+
+    urls = [
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/logic.py]: Not the most important point (not a blocker), but the list seems more confusing than simply variables. E.g.:

``` python3
url_350 = cloudinary + "canonical/image/fetch/q_auto,f_auto,w_350/\g<url>"
url_650 = cloudinary + "canonical/image/fetch/q_auto,f_auto,w_650/\g<url>"
url_1300 = cloudinary + "canonical/image/fetch/q_auto,f_auto,w_1300/\g<url>"
url_1950 = cloudinary + "canonical/image/fetch/q_auto,f_auto,w_1950/\g<url>"
```

(I may have written this initially, so I'm probably arguing with myself)

Comment (nottrobin) [reply canonicalwebteam/blog_extension/logic.py]: Or how about:

``` python3
    image_prefix = cloudinary + "canonical/image/fetch/q_auto,f_auto"

    image_match = (
        r'<img(?P<prefix>[^>]*) src="(?P<url>[^"]+)"(?P<suffix>[^>]*)>'
    )
    replacement = (
        "<img\g<prefix>"
        f' decoding="async"'
        f' src="{image_prefix},w_350"'
        f' srcset={image_prefix},w_350 350w, {image_prefix},w_650 650w,'
        f' {image_prefix},w_1300 1300w, {image_prefix},w_1950 1950w"'
        f' sizes="(max-width: 400px) 350w, 650px"'
        "\g<suffix>>"
    )
```

File: canonicalwebteam/blog_extension/api.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    if not response.ok:
+        return None
+
+    return process_response(response)
+
+
+def get_feed(tag):
+    """"""
+    response = api_session.get(
+        "https://admin.insights.ubuntu.com/?tag={}&feed=rss".format(tag)
```
[the remark this answers is not in the document]
Comment (nottrobin) [reply canonicalwebteam/blog_extension/api.py]: unfortunately not. I think this module needs to support not only < 3.6, but also <3. fstrings are too modern. We don't need to add python2 functionality yet, but we certainly shouldn't introduce dependency on >= 3.6.

File: canonicalwebteam/blog_extension/api.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
+import os
```
[the remark this answers is not in the document]
Comment (nottrobin) [reply canonicalwebteam/blog_extension/api.py]: :+1:

File: canonicalwebteam/blog_extension/api.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
```
Comment (nottrobin) [inline canonicalwebteam/blog_extension/api.py]: As discussed, could we rename `blog_extension` to `blog`?

File: canonicalwebteam/blog_extension/logic.py (hunk 6 above)
Comment (nottrobin) [reply canonicalwebteam/blog_extension/logic.py]: Okay issue filed for this: https://github.com/canonical-webteam/blog-flask-extension/issues/3

Resolving this thread.

File: canonicalwebteam/blog_extension/views.py (hunk 4 above)
Comment (nottrobin) [reply canonicalwebteam/blog_extension/views.py]: So this doesn't take too long to be merged, I've agreed with Toto that we defer anything more than core fixes to future issues. 

So I've filed an issue for this point - https://github.com/canonical-webteam/blog-flask-extension/issues/4 - and I'll resolve this conversation. Hope that's okay.

File: canonicalwebteam/blog_extension/logic.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        # split at the last 3 characters
+        raw_article_start = raw_article[:-3]
+        raw_article_end = raw_article[-3:]
+        # for the last 3 characters replace any part of […]
+        raw_article_end = raw_article_end.replace("[", "")
+        raw_article_end = raw_article_end.replace("…", "")
+        raw_article_end = raw_article_end.replace("]", "")
+        # join it back up
+        article["excerpt"]["raw"] = "".join(
+            [raw_article_start, raw_article_end, " […]"]
```
[the remark this answers is not in the document]
Comment (nottrobin) [reply canonicalwebteam/blog_extension/logic.py]: Again I've filed this as an issue so we can explore it down the line, to get this merged quickly:

https://github.com/canonical-webteam/blog-flask-extension/issues/5

Resolving.

File: canonicalwebteam/blog_extension/logic.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+
+def strip_excerpt(raw_html):
+    """Remove tags from a html string
+
+    :param raw_html: The HTML to strip
+
+    :returns: The stripped string
+    """
+    cleanr = re.compile("<.*?>")
```
[the remark this answers is not in the document]
Comment (nottrobin) [reply canonicalwebteam/blog_extension/logic.py]: Moved to an issue: https://github.com/canonical-webteam/blog-flask-extension/issues/6

Resolving

File: canonicalwebteam/blog_extension/logic.py (hunk 2 above)
Comment (nottrobin) [reply canonicalwebteam/blog_extension/logic.py]: Moved to an issue: https://github.com/canonical-webteam/blog-flask-extension/issues/7

File: canonicalwebteam/blog_extension/logic.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tbille) [reply canonicalwebteam/blog_extension/logic.py]: this was for `clean_regex` I am renaming it now

File: canonicalwebteam/blog_extension/views.py (hunk 4 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/views.py]: this is my mistake, removing it now

File: canonicalwebteam/blog_extension/logic.py (hunk 6 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/logic.py]: as discussed, this should be an issue for now

File: canonicalwebteam/blog_extension/logic.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    ):
+        article["content"]["rendered"] = replace_images_with_cloudinary(
+            article["content"]["rendered"]
+        )
+
+    return article
+
+
+def change_url(feed, host):
+    """Change insights urls to snapcraft.io/blog
```
[the remark this answers is not in the document]
Comment (tbille) [reply canonicalwebteam/blog_extension/logic.py]: done

File: canonicalwebteam/blog_extension/api.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tbille) [reply canonicalwebteam/blog_extension/api.py]: done

File: canonicalwebteam/blog_extension/api.py (hunk 9 above)
Comment (tbille) [reply canonicalwebteam/blog_extension/api.py]: done

**Pull request merged**